### PR TITLE
25706/chatbot cypress

### PR DESCRIPTION
**Pull request by CBonade**
## Description
https://app.zenhub.com/workspaces/vsp-5cedc9cce6e3335dc5a49fc4/issues/department-of-veterans-affairs/va.gov-team/25706

Tests rewritten for the coronavirus chatbot.

Test 1 - Successful load:
This test was currently being skipped in the Nightwatch testing.  As this original test was set up to use a real API, Cypress was having some problems with this as it gets unavoidable (at least in the research I did) cross-origin security problems.  As opposed to normal Chrome where you can disable the security in the browser, I was unable to do this inside Cypress' testing (even the config option for cypress to disable `chromeWebSecurity` didn't seem to do the trick).  I also, after looking at the react render code, attempted to first fetch a token and store it in localStorage (as well as mocking a token and storing it), only to have the error state load anyway.  That said, I've rewritten the test to the best of my ability without being able to test it all the way through.

Test 2 - Error state: 
This test has been rewritten, tested and all passes without issue.

## Testing done
Test 1 does not pass due to not being able to load the success state over the API.  Test is being skipped in the same manner as it was in the Nightwatch testing suite.

Test 2 has been run in Chrome and passes on all accounts

## Screenshots
Test 1 skipped:

image

Test 2: 

image

## Acceptance criteria
- [ ]

## Definition of done
- [ ] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/applications/coronavirus-chatbot/tests/01-buttons.cypress.spec.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      data: chatbotResponse,
+    }).as('getChatbot');
+
+    cy.visit(coronavirusChatbotPath);
+
+    cy.wait('@getChatbot');
+
+    cy.get('body', { timeout: Timeouts.verySlow }).should('be.visible');
+    cy.title().should('eq', 'VA coronavirus chatbot | Veterans Affairs');
+    cy.log('@getChatbot');
```
Comment (ddzz) [inline src/applications/coronavirus-chatbot/tests/01-buttons.cypress.spec.js:28]: Can this line be taken out?

File: src/applications/coronavirus-chatbot/tests/00-load-error.cypress.spec.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+const Timeouts = require('platform/testing/e2e/timeouts');
+
+const loadError = 'div.usa-alert.usa-alert-error';
+const loadErrorHeading = 'h3.usa-alert-heading';
+const coronavirusChatbotPath = '/coronavirus-chatbot/';
+
+describe('Chatbot Load Error', () => {
+  it('Shows proper error message', () => {
+    cy.intercept('/v0/feature_toggles?&cookie_id=*', {
```
Comment (ddzz) [inline src/applications/coronavirus-chatbot/tests/00-load-error.cypress.spec.js:9]: Is `?&cookie_id=*` needed here?

File: src/applications/coronavirus-chatbot/tests/01-buttons.cypress.spec.js (hunk 1 above)
Comment (karlbrown-va) [reply src/applications/coronavirus-chatbot/tests/01-buttons.cypress.spec.js:28]: which line?

File: src/applications/coronavirus-chatbot/tests/01-buttons.cypress.spec.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    cy.get(vaAvatarMatcher, { timeout: 45000 }).should('be.visible');
```
Comment (ddzz) [inline src/applications/coronavirus-chatbot/tests/01-buttons.cypress.spec.js:29]: Can `Timeouts` be used here instead of a hard-coded value?

**Comment by karlbrown-va** [comment]
If I recall correctly the original intent of this was test an e2e smoke test. If we don't have any test that validates the chatbot loads it's not helping much. I don't know the story of how the original test was disabled (I don't recall if it ever worked, though i *think* it did). Do we have other options for having a test that validates that (with latest changes) - the chatbot continues to show up?

File: src/applications/coronavirus-chatbot/tests/01-buttons.cypress.spec.js (hunk 1 above)
Comment (ddzz) [reply src/applications/coronavirus-chatbot/tests/01-buttons.cypress.spec.js:28]: L28, the `cy.log()`.

Comment (CBonade) [reply src/applications/coronavirus-chatbot/tests/01-buttons.cypress.spec.js]: Good call, removed!

File: src/applications/coronavirus-chatbot/tests/00-load-error.cypress.spec.js (hunk 2 above)
Comment (CBonade) [reply src/applications/coronavirus-chatbot/tests/00-load-error.cypress.spec.js]: Interesting - the first time I tried this without it, it failed the request.  Trying it now without it, it seems to have worked.  Possible I had the formatting incorrect the first time.  I've removed this in both files and left just the base `/v0/feature_toggles`

**Comment by CBonade** [comment]
In regards to the intention of our update/ticket here, it was strictly to port any existing Nightwatch tests over to Cypress.  As far as the intentions of the test or a better way to approach it, I fear I lack the familiarity with the chatbot itself to make that adjustment and might be more suited for someone who has worked on the chatbot.  The second test, checking the error state, has been confirmed functional, however.

**Comment by CBonade** [comment]
Discussed options with the team.  We'd be happy to take on the task of finding a way to better test the chatbot, however we'd just prefer to keep it as a separate issue from the conversion effort that this PR was based around.  We will create a separate issue to write a new test for that and find an owner that's more familiar with the chatbot to take that on.  In the meantime, at least the fail state test here is good to go and minimally, the success test is in the same state it was in prior.  Additionally, we are not at a stage yet of removing Nightwatch tests (that's next phase), so we are not losing any coverage in the interim.

**Comment by pjhill** [comment]
> If I recall correctly the original intent of this was test an e2e smoke test. If we don't have any test that validates the chatbot loads it's not helping much. I don't know the story of how the original test was disabled (I don't recall if it ever worked, though i _think_ it did). Do we have other options for having a test that validates that (with latest changes) - the chatbot continues to show up?

It doesn't look like the test has been updated since the original PR was merged. Per the conversations in the original PR, it looks like mocks won't work as the chatbot requires a valid API token from Microsoft in order to render correctly. That'll pose a challenge. We may revisit this issue in the near future, but for now we are just looking to reproduce existing behaviors in Cypress instead of Nightwatch, so we can close out Nightwatch.

**Review by pjhill — APPROVED** [review]
LGTM

**Comment by karlbrown-va** [comment]
> Discussed options with the team. We'd be happy to take on the task of finding a way to better test the chatbot, however we'd just prefer to keep it as a separate issue from the conversion effort that this PR was based around. We will create a separate issue to write a new test for that and find an owner that's more familiar with the chatbot to take that on. In the meantime, at least the fail state test here is good to go and minimally, the success test is in the same state it was in prior. Additionally, we are not at a stage yet of removing Nightwatch tests (that's next phase), so we are not losing any coverage in the interim.

Hi, thanks for this, that seems quite reasonable; and indeed I may be misremembering and perhaps this never worked properly. Please tag me on the new issue and we can have our team add some thoughts and help put this together.

**Review by jgacek — APPROVED** [review]

**Comment by pjhill** [comment]
Resolves [25706](https://github.com/department-of-veterans-affairs/va.gov-team/issues/25706)

**Pull request merged**